docs/sphinx/user_guide/feature/resource.rst

```diff
@@ -0,0 +1,251 @@
+Cuda resource contains a ``cudaStream_t`` value with an associated get method. See the 
+individual functionality for each resource in ``raja/tpl/camp/include/resource/``.
+
+
+------------
+Type-Erasure
+------------
+
+Resources can be declared in two formats: An erased resource, and a concrete resource. When 
+declared on the same backend, their underlying runtime functionality is the same. An 
+erased resource allows a user the ability to change the resourse backend at runtime. 
+
+Concrete Cuda resource::
+
+    RAJA::resources::Cuda my_cuda_res;
+
+Erased resource::
+
+    if (use_gpu)
+      RAJA::resources::Resource my_res{RAJA::resources::Cuda()};
+    else
+      RAJA::resources::Resource my_res{RAJA::resources::Host()};
+
+
+Memory allocation on resources::
+
+    int* a1 = my_cuda_res.allocate<int>(ARRAY_SIZE);
+    int* a2 = my_res.allocate<int>(ARRAY_SIZE);
+
+If ``use_gpu`` is ``true``, the underlying type of ``my_res`` is a Cuda resource, therefore ``a1`` and ``a2`` will both be allocated on the GPU. If ``use_gpu`` is ``false``, then only ``a1`` is allocated on the GPU, and ``a2`` is allocated on the host.
+
+
+------
+Forall
+------
+
+A resource is an optional argument to a ``RAJA::forall`` call. It is passed as the first argument when the forall is templated on the execution policy. When the execution policy is passed by value, the resource is passed as the second argument. This maintains the order of conditions that can be passed to a ``RAJA::forall`` call::
+
+    RAJA::forall<ExecPol>(my_cuda_res, .... )
+
+or::
+
+    RAJA::forall(ExecPol(), my_cuda_res, .... )
+
+
+When specifying a Cuda/Hip resource the RAJA kernel is executed aynchronously on a stream.
+Currently Cuda/Hip are the only Resources that enable asynchronous threading with a ``RAJA::forall``.
+All other calls are defaulted to using the ``Host`` resource until further support is 
+added.
+
+The Resource type that is passed to a ``RAJA::forall`` call must be a concrete type. This is to
+allow compile time assertion that the resource is compatible with the given execution policy. For
+example::
+    
+    using ExecPol = RAJA::cuda_exec_async<BLOCK_SIZE>;
+    RAJA::resources::Cuda my_cuda_res;
+    RAJA::resources::Resource my_res{RAJA::resources::Cuda()};
+    RAJA::resources::Host my_host_res;
+
+    RAJA::forall<ExecPol>(my_cuda_res, .... ) // Compiles.
+    RAJA::forall<ExecPol>(my_res, .... ) // Compilation Error. Not Concrete.
+    RAJA::forall<ExecPol>(my_host_res, .... ) // Compilation Error. Mismatched Resource and Exec Policy.
+
+Below is a list of the current concrete resource type and their execution policy suport.
+
+ ======== ==============================
+ Resource Policy Type
+ ======== ==============================
+ Cuda     cuda_exec
+
+          cuda_exec_async
+
+ Hip      hip_exec
+
+          hip_exec_async
+
+ Omp*     omp_target_parallel_for_exec
+          omp_target_parallel_for_exec_n
```

Comment:
Has this been tested?

Reply:
Omp target does not work as explained in the note below this.